route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -0,0 +1,176 @@
+ * the License.
+ */
+
+#include <ros/ros.h>
+#include <string>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include "route_following_plugin.h"
+
+
+namespace route_following_plugin
+{
+    RouteFollowingPlugin::RouteFollowingPlugin() : mvr_duration_(16.0), current_speed_(0.0) { }
+
+    void RouteFollowingPlugin::initialize()
+    {
+        
+        nh_.reset(new ros::CARMANodeHandle());
+        pnh_.reset(new ros::CARMANodeHandle("~"));
+        
+        plan_maneuver_srv_ = nh_->advertiseService("plugins/RouteFollowingPlugin/plan_maneuvers", &RouteFollowingPlugin::plan_maneuver_cb, this);
+                
+        plugin_discovery_pub_ = nh_->advertise<cav_msgs::Plugin>("plugin_discovery", 1);
+        plugin_discovery_msg_.name = "RouteFollowingPlugin";
+        plugin_discovery_msg_.versionId = "v1.0";
+        plugin_discovery_msg_.available = true;
+        plugin_discovery_msg_.activated = false;
+        plugin_discovery_msg_.type = cav_msgs::Plugin::STRATEGIC;
+        plugin_discovery_msg_.capability = "strategic_plan/plan_maneuvers";
+
+        pose_sub_ = nh_->subscribe("current_pose", 1, &RouteFollowingPlugin::pose_cb, this);
+        twist_sub_ = nh_->subscribe("current_velocity", 1, &RouteFollowingPlugin::twist_cd, this);
+        
+        pnh_->param<double>("minimal_maneuver_duration", mvr_duration_, 16.0);
+
+        wml_.reset(new carma_wm::WMListener());
+        // set world model point form wm listener
+        wm_ = wml_->getWorldModel();
+
+        ros::CARMANodeHandle::setSpinCallback([this]()
+        {
+            plugin_discovery_pub_.publish(plugin_discovery_msg_);
+            return true;
+        });
+    }
+
+    void RouteFollowingPlugin::run()
+    {
+        initialize();
+        ros::CARMANodeHandle::spin();
+    }
+
+    bool RouteFollowingPlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp)
+    {        
+
+        lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 1);
+        if(current_lanelets.size() == 0)
+        {
+            ROS_WARN_STREAM("Cannot find any lanelet in map!");
+            return true;
+        }
+        auto current_lanelet = current_lanelets[0];
+        auto shortest_path = wm_->getRoute()->shortestPath();
+        double current_progress = wm_->routeTrackPos(current_loc).downtrack;
+        double speed_progress = current_speed_;
+        double total_maneuver_length = current_progress + mvr_duration_ * RouteFollowingPlugin::TWENTY_FIVE_MPH_IN_MS;
+        int last_lanelet_index = findLaneletIndexFromPath(current_lanelet.second.id(), shortest_path);
+        if(last_lanelet_index == -1)
+        {
+            ROS_WARN_STREAM("Cannot find current lanelet in shortest path!");
```

Comment:
This is the case where we'd expect `RouteManager` or someone else to throw a left route error, correct? Just making sure I understand.

Reply:
If we stay on the route, it will not be trigger. Otherwise, we expect route manager to flag left_route first.